### ISSUE TITLE
[Error] Can't install Snaps, Udev Permission Error

### ISSUE THREAD

**Issue by japzone1** [89luca89/distrobox#324]
**Describe the bug**
My end goal is to try and get Snaps working, but my current progress is stopped by a DPKG error when installing Fuse. Haven't been able to find a log of the error(DPKG just says there was an error), but from what I've read online and what I've observed trying other things, it seems to be some kind of permission error involving either Udev or /dev/fuse.

**To Reproduce**
1. Setup Distrobox without root: `curl -s https://raw.githubusercontent.com/89luca89/distrobox/main/install | sh -s -- --prefix ~/.local`
2. Create container that supports Init: `distrobox create -i docker.io/nimmis/ubuntu:latest --init --name ubuntu-init
`
3. Enter container and install snapd, fuse, and attempt to install a snap.
```
distrobox enter ubuntu-init
sudo apt install snapd fuse
sudo systemctl start snapd.socket
sudo snap install nordpass
```

**Expected behavior**
Snap installs.

**Logs**
Run the commands with `--verbose` and post the log here as a file upload
Attach also the output of `podman logs` or `docker logs`, possibly with `--latest` flag
[apt-snap.log](https://github.com/89luca89/distrobox/files/8913375/apt-snap.log)
[podman.log](https://github.com/89luca89/distrobox/files/8913376/podman.log)

**Desktop (please complete the following information):**
 - Are you using podman or docker? Podman
 - Which version or podman or docker? podman version 3.4.4
 - Which version of distrobox? distrobox: 1.3.1

**Additional context**
neofetch:
```
              .,,,,.                  deck@steamdeck 
        .,'onNMMMMMNNnn',.            -------------- 
     .'oNMANKMMMMMMMMMMMNNn'.         OS: SteamOS Holo x86_64 
   .'ANMMMMMMMXKNNWWWPFFWNNMNn.       Host: Jupiter 1 
  ;NNMMMMMMMMMMNWW'' ,.., 'WMMM,      Kernel: 5.13.0-valve15-1-neptune-02197-gf6ec7ad3762a 
 ;NMMMMV+##+VNWWW' .+;'':+, 'WMW,     Uptime: 1 day, 6 hours, 47 mins 
,VNNWP+######+WW,  +:    :+, +MMM,    Packages: 905 (pacman), 28 (flatpak) 
'+#############,   +.    ,+' +NMMM    Shell: bash 5.1.16 
  '*#########*'     '*,,*' .+NMMMM.   Resolution: 1280x800 
     `'*###*'          ,.,;###+WNM,   DE: Plasma 5.23.5 
         .,;;,      .;##########+W    WM: KWin 
,',.         ';  ,+##############'    Theme: Vapor [Plasma], Breeze [GTK2/3] 
 '###+. :,. .,; ,###############'     Icons: [Plasma], breeze-dark [GTK2/3] 
  '####.. `'' .,###############'      Terminal: konsole 
    '#####+++################'        CPU: AMD Custom APU 0405 (8) @ 2.800GHz 
      '*##################*'          GPU: AMD ATI 04:00.0 VanGogh 
         ''*##########*''             Memory: 6174MiB / 14838MiB 
              ''''''
```

**Comment by 89luca89** [comment]
Hi @japzone1 

This is mostly related to the use of `fuse` and being in a rootless container, it much likely won't work

You could try using `--root` when creating the container, but be aware that ROOTFUL containers can modify your system, and can interfere with thing, especially if you launch with `--init`

I'm working on "taming" a bit systemd in the `init` so that it won't clash with host's instance, but it's not as safe as using rootless containers obviously

 Sadly for now snaps inside containers (being docker or podman in general) it's quite tricky

**Comment by 89luca89** [comment]
I've just pushed commit https://github.com/89luca89/distrobox/commit/881e60b6bd51c6caee577e87c78741c3ae6a30ad
that improves the use of rootful containers with systemd/inits

Worth giving a try perhaps

**Comment by 89luca89** [comment]
I can confirm that with --root, and last commit, snapd works inside the container :+1: leaving to you to test it, but I don't think it will be possible to use snapd with a rootless container

**Comment by japzone1** [comment]
> I can confirm that with --root, and last commit, snapd works inside the container +1 leaving to you to test it, but I don't think it will be possible to use snapd with a rootless container
@89luca89 

Cool, I'll give it a shot.

-----

Well, I'm getting a storage space error for some reason. I double checked and none of my partitions are full, and this doesn't happen when I'm not trying to use root mode. I remember seeing this error before when I was messing around with other things, so it's not a new thing, but didn't think about it then. Is it because I installed Distrobox without root?

```
(deck@steamdeck ~)$ distrobox create -i docker.io/nimmis/ubuntu:latest --init --name ubuntu-test --root
Image docker.io/nimmis/ubuntu:latest not found.
Do you want to pull the image now? [Y/n]: y
Trying to pull docker.io/nimmis/ubuntu:latest...
Getting image source signatures
Copying blob 4d32b49e2995 skipped: already exists  
Copying blob 1bb2b2c5efdf skipped: already exists  
Copying blob 4f4fb700ef54 done  
Copying blob 7bccc863f2da done  
Error: writing blob: adding layer with blob "sha256:7bccc863f2dac9c245f9de5adabdb58ddaa053d8a93604f5fbe0effc268516f6": Error processing tar file(exit status 1): write /usr/lib/x86_64-linux-gnu/libicudata.so.66.1: no space left on device

An error occurred

An error occurred
(125)(deck@steamdeck ~)$ df -h
Filesystem      Size  Used Avail Use% Mounted on
devtmpfs        7.3G     0  7.3G   0% /dev
tmpfs           7.3G  491M  6.8G   7% /dev/shm
tmpfs           2.9G  9.8M  2.9G   1% /run
/dev/nvme0n1p4  5.0G  3.3G  1.5G  70% /
/dev/nvme0n1p6  230M  106M  108M  50% /var
overlay         230M  106M  108M  50% /etc
/dev/nvme0n1p8  227G  201G   27G  89% /home
tmpfs           7.3G   19M  7.3G   1% /tmp
tmpfs           1.5G  260K  1.5G   1% /run/user/1000
/dev/mmcblk0p1  469G  450G   19G  96% /run/media/mmcblk0p1
```

**Comment by 89luca89** [comment]
This is much likely device-dependent, in root mode podman uses `/var` and you only have 100MB left on it
So that's most likely why

Better refer to podman's manual on how to configure it to use another directory :+1:

**Comment by 89luca89** [comment]
@japzone1 any news on this?

Storage errors are out of scope on this, as it's a podman/partitions issue
so if it's ok I'll close this as snaps are working with the rootful setup I've suggested

**Comment by japzone1** [comment]
@89luca89 Sorry, I've been busy with work. I've got free time tomorrow though so I'll be testing it then.

**Comment by japzone1** [comment]
RL sucks.

@89luca89 
Anyway, got the storage issue resolved, and managed to build the container, but launching it has lead to a different problem.

I ran:
```
distrobox create -i docker.io/nimmis/ubuntu:latest --init --name ubuntu-init --root
distrobox enter --root ubuntu-init
```

Everything goes fine, until right when I get dropped into the container's shell and suddenly my Steam Deck blackscreens. I think XServer is crashing or something, since I can still remote into a shell on the SD from my phone/PC, but the GUI is dead for the most part with my remote software not being able to pull it up. Thankfully no permanent damage is done since a hard reboot works fine. I guess once the Ubuntu container tries to load Init, it tries to mess with the XServer for some reason? Not sure. Did you use a different container when testing on your end?
[distrobox.log](https://github.com/89luca89/distrobox/files/8972082/distrobox.log)
[distrobox-podman.log](https://github.com/89luca89/distrobox/files/8972083/distrobox-podman.log)

**Comment by 89luca89** [comment]
@japzone1 

did you use latest 1.3.2 version from `git`? there were a lot of updates to improve this type of problems on rootful init containers
(you will need to stop and start again the container when you update)

**Comment by japzone1** [comment]
@89luca89 I'm pretty sure I was using the latest git, but I'll try again tomorrow

**Comment by 89luca89** [comment]
> @89luca89 I'm pretty sure I was using the latest git, but I'll try again tomorrow

Ok let me know, I remember having some issue similar (gnome reloading) before the commits I've pushed, that's why I asked

**Comment by japzone1** [comment]
> > @89luca89 I'm pretty sure I was using the latest git, but I'll try again tomorrow
> 
> Ok let me know, I remember having some issue similar (gnome reloading) before the commits I've pushed, that's why I asked

Sorry for the late reply. I had a bad week.

Anyway, to be sure I was on the latest version I uninstalled distrobox and then installed it again using:
`curl -s https://raw.githubusercontent.com/89luca89/distrobox/main/install | sh -s -- --next --prefix ~/.local`

I then proceeded to repeat my [previous steps](https://github.com/89luca89/distrobox/issues/324#issuecomment-1165026129), but unfortunately I got the same result. Right when distrobox drops me into the container's shell the window server crashes, giving me a black screen that requires a reboot to fix.

The logs look pretty much the same. Are there any other logs I could grab that would help?

**Comment by 89luca89** [comment]
Hey @japzone1 
seems strange, as [this commit](https://github.com/89luca89/distrobox/commit/b79564659646368253c6f844641bc41a6718f0ec) solved the problem, as confirmed also in #344 

To check if you're using a version that has fixed the issue, just go in to the container and check if `/dev/console` is a device, or is an empty file
Because the commit introduced a fix that would overwrite `/dev/console` with an empty file, as that was triggering the UI session to restart

**Comment by 89luca89** [comment]
> At this point /dev/console is a device.

this is a problem, in the `init` it's replaced by an empty file BEFORE starting systemd at all, so this shouldn't happen at all, when you're dropped inside the container the file should already be replaced 

Can you attach here the output of

`sudo podman inspect ubuntu-init`

**Comment by japzone1** [comment]
@89luca89 
> Can you attach here the output of
> 
> `sudo podman inspect ubuntu-init`

[podman-inspect-ubnutu-init.log](https://github.com/89luca89/distrobox/files/9060471/podman-inspect-ubnutu-init.log)

**Comment by 89luca89** [comment]
mmh Seems all ok

So just to recap
If you create a new container with the upgraded version of distrobox (git)

- `distrobox create --root -i docker.io/nimmis/ubuntu:latest --init --name ubuntu-init `
- `distrobox enter --root ubuntu-init`

Without doing any other setup (snap or other), does it crash the graphical session?

**Comment by 89luca89** [comment]
Yep, let's proceed this way:

```sh
distrobox stop --root ubuntu-init
distrobox enter --root ubuntu-init
```

while you do `enter`, in another tab you can follow with

`sudo podman logs -f ubuntu-init`

this podman logs command will follow the initialization the container step by step, and will see if something is wrong

**Comment by 89luca89** [comment]
To do it somplete, you could do

`sudo podman logs -f ubuntu-init 2>&1 > output.log`

so this way it will be complete and no need to copy-paste

**Comment by japzone1** [comment]
> To do it somplete, you could do
> 
> `sudo podman logs -f ubuntu-init 2>&1 > output.log`
> 
> so this way it will be complete and no need to copy-paste

Didn't work. I think what's happening is that for some reason the terminal is displaying the list of commands that were run, while the `> output.txt` is only displaying the output of those commands. No idea why `podman logs` works like that though.

**Comment by japzone1** [comment]
> Sorry I've inverted the redirects, it should be
> 
> `sudo podman logs -f ubuntu-init > output.log 2>&1 `
> 
> and should work

Ah, that worked. So that was the magic I've been trying to find with Google.

[podman-logs-ubuntu-init3.log](https://github.com/89luca89/distrobox/files/9065558/podman-logs-ubuntu-init3.log)

**Comment by japzone1** [comment]
@89luca89 
Is `curl -s https://raw.githubusercontent.com/89luca89/distrobox/main/install | sh -s -- --next --prefix ~/.local` actually getting me the latest version? As far as I can tell, the files at least seem the same as the latest in the git, but maybe I'm missing something?

**Comment by 89luca89** [comment]
yea seems it is the right version indeed, and it's doing what is intended to do really

can you try with some other distro image? like an almalinux or redhat systemd image? To see if there is something else in play at this point

**Comment by japzone1** [comment]
Ok, I used:
```
distrobox create -i docker.io/almalinux/9-init:latest --init --name alma --root
distrobox enter --root alma
```
And the GUI doesn't crash. But I'm not familiar with Red Hat based distros so this a bit weird to use for me.

I followed Snapcraft.io's CentOS instructions to get Snap installed and Snapd and everything seems to be working, 
```
(deck@alma ~)$ snap --version
snap       2.56.2-1.el9
snapd      2.56.2-1.el9
series     16
almalinux  9.0
kernel     5.13.0-valve15-1-neptune-02197-gf6ec7ad3762a
```
but I run into issues once I install snaps.

To keep things simple, I'm testing using Kapman and Kmines, as they serve as simple GUI apps without anything too crazy going on.  So I try to install Kapman, and while it technically installs, this is where issues crop up:
```
(deck@alma ~)$ sudo snap install kapman
error: cannot perform the following tasks:
- Mount snap "kde-frameworks-5-qt-5-15-3-core20" (8) (systemctl command [start var-lib-snapd-snap-kde\x2dframeworks\x2d5\x2dqt\x2d5\x2d15\x2d3\x2dcore20-8.mount] failed with exit status 1: Job failed. See "journalctl -xe" for details.)
- Mount snap "gtk-common-themes" (1535) (systemctl command [start var-lib-snapd-snap-gtk\x2dcommon\x2dthemes-1535.mount] failed with exit status 1: Job failed. See "journalctl -xe" for details.)
```
Just for fun I try to run the app and get this error:
```
(deck@alma ~)$ kapman
internal error, please report: running "kapman" failed: cannot create transient scope: DBus error "org.freedesktop.DBus.Error.UnixProcessIdUnknown": [Process with ID 4944 does not exist.]
```
I thought maybe this is just because there's no GUI inside the container, so I try to export the game with Distrobox and get this:
```
(deck@alma ~)$ distrobox-export --app kapman
Error: cannot find any desktop files.
Error: trying to export a non-installed application.

An error occurred
```

Not sure where to go from here.

**Comment by japzone1** [comment]
Also tried installing Kmines via Yum instead and then exporting, but that won't launch in SteamOS either:
```
(deck@steamdeck ~)$ /home/deck/.local/bin/distrobox-enter --root -n alma -- '  kmines -q windowtitle  %c'
[sudo] password for deck:
Authorization required, but no authorization protocol specified
qt.qpa.xcb: could not connect to display :0
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx, xcb.

An error occurred
```

**Comment by 89luca89** [comment]
So, two differen issues here:



> Also tried installing Kmines via Yum instead and then exporting, but that won't launch in SteamOS either:
> 
> ```
> (deck@steamdeck ~)$ /home/deck/.local/bin/distrobox-enter --root -n alma -- '  kmines -q windowtitle  %c'
> [sudo] password for deck:
> Authorization required, but no authorization protocol specified
> qt.qpa.xcb: could not connect to display :0
> qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
> This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
> 
> Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx, xcb.
> 
> An error occurred
> ```

For this you need to refer to the xhost command needed, as explained in the documentation [HERE](https://github.com/89luca89/distrobox/blob/main/docs/compatibility.md#compatibility-notes)



> Just for fun I try to run the app and get this error:
> 
> ```
> (deck@alma ~)$ kapman
> internal error, please report: running "kapman" failed: cannot create transient scope: DBus error "org.freedesktop.DBus.Error.UnixProcessIdUnknown": [Process with ID 4944 does not exist.]
> ```

This is because to have an init system INSIDE the container, we cannot share the same PID namespace as the host.
So, there is an incompatibility in dbus in this case, trying to find a pid that the container cannot see


> I thought maybe this is just because there's no GUI inside the container, so I try to export the game with Distrobox and get this:
> 
> ```
> (deck@alma ~)$ distrobox-export --app kapman
> Error: cannot find any desktop files.
> Error: trying to export a non-installed application.
> 
> An error occurred
> ```
> 
> Not sure where to go from here.

This is instead because to `export`, distrobox uses the app's desktop file, which needs to be in a specific location, and snaps are not supported now

---

All in all I see that snaps inside a container are quite difficult and I don't think I'll ever say that snap-inside-distrobox is something supported :stuck_out_tongue:

**Comment by 89luca89** [comment]
@japzone1 at this point I think I'll just add some warnings of snaps unsupported inside a distrobox, and point to this issue
I don't think it is easily fixable, and it's not really a use case I'm personally interested in covering (if someone else wants, it's warmly welcomed :smile: )

**Comment by IPlayZed** [comment]
Just for curiosity, do you think making a feature request add the Snap project to support running inside rootless containers is something worth the time?
